### PR TITLE
Use the /var/lib/calico/nodename file if it exists

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

In combination with https://github.com/projectcalico/calico/pull/1722, will make it much harder to accidentally end up with mismatched node names between the two components.

Fixes https://github.com/projectcalico/calico/issues/1093

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
The Calico CNI plugin by default expects the /var/lib/calico/nodename file to be created by calico/node. To disable this feature, set `nodename_file_optional: true` in your CNI network configuration.
```

### CODE REVIEW

File: calico.go [hunk 1]
```diff
@@ -56,8 +54,16 @@ func cmdAdd(args *skel.CmdArgs) error {
...
 
+	if conf.RequireNodenameFile {
+		// Configured to use the nodename file - don't start until it exists.
+		if _, err := os.Stat("/var/lib/calico/nodename"); err != nil {
+			return fmt.Errorf("error reading /var/lib/calico/nodename: %s: is calico/node running?", err)
+		}
+		logrus.Debug("/var/lib/calico/nodename exists")
+	}
+
 	// Allow the nodename to be overridden by the network config
```
Comment (ozdanborne) [inline calico.go]: Now that nodename is being instantiated here, I don't think this comment is right

File: utils/utils.go [hunk 2]
```diff
@@ -57,12 +58,34 @@ func DetermineNodename(conf types.NetConf) string {
...
+		logrus.Debugf("Read node name from CNI conf: %s", conf.Nodename)
 		nodename = conf.Nodename
 	}
+	logrus.Debugf("Using node name %s", nodename)
 	return nodename
 }
 
+// nodenameFromFile reads the /var/lib/calico/nodename file if it exists and
+// returns the nodename within.
+func nodenameFromFile() string {
```
Comment (ozdanborne) [inline utils/utils.go]: A nodeNameFile that exists but is empty takes an interesting path through this code, but I suppose it is ok...

File: utils/utils.go [hunk 3]
```diff
@@ -57,12 +58,34 @@ func DetermineNodename(conf types.NetConf) string {
 		nodename = conf.Hostname
 		logrus.Warn("Configuration option 'hostname' is deprecated, use 'nodename' instead.")
 	}
+	if nff := nodenameFromFile(); nff != "" {
+		logrus.Debugf("Read node name from file: %s", nff)
+		nodename = nff
+	}
 	if conf.Nodename != "" {
```
Comment (ozdanborne) [inline utils/utils.go]: Should this be an `else if`? If multiple methods are provided (albeit an unusual edge case), the lowest priority would take precedence.

File: calico.go (hunk 1 above)
Comment (caseydavenport) [reply calico.go]: I don't think the comment was ever right.

File: utils/utils.go (hunk 3 above)
Comment (caseydavenport) [reply utils/utils.go]: The way the code is structured now is acting as designed - the `conf.Nodename` takes priority over the `/var/lib/calico/nodename` file.

File: utils/utils.go (hunk 2 above)
Comment (caseydavenport) [reply utils/utils.go]: Hm, yeah. I suspect we probably want to error in that case.

Comment (caseydavenport) [reply utils/utils.go]: Actually, let's leave this as it is for now. I think the code will do the right thing and I'm not sure how we'd end up with an empty file anyway.

**Pull request merged**